### PR TITLE
validate: ganesha validation is too strict

### PR DESCRIPTION
**Pull request by jan--f**
Ganesha validation should only fail on missing mds or rgw node if
ganesha is actually deployed.
Also adds some refactoring to simplify pillar iteration.
Fixes: #128 

Signed-off-by: Jan Fajerski <jfajerski@suse.com>

### CODE REVIEW

File: srv/modules/runners/validate.py [hunk 1]
```diff
@@ -278,37 +278,41 @@ def storage(self):
...
-                    ganesha_roles = list(set(self.data[node].get("roles")) &
-                                        set(self.data[node].get("ganesha_configurations")))
+        for node, data in self.data.items():
+            if ('roles' in data):
+                if('ganesha_configurations' in data):
+                    ganesha_roles = list(set(data.get("roles")) &
+                                        set(data.get("ganesha_configurations")))
                     if len(ganesha_roles) > 1:
                         msg = "minion {} has {} roles. Only one permitted".format(node, ganesha_roles)
                         self.errors.setdefault('ganesha',[]).append(msg)
```
Comment (supriti) [inline srv/modules/runners/validate.py]: We can set role_ganesha = True if len(ganesa_roles) is not greater than 1.

File: srv/modules/runners/validate.py [hunk 2]
```diff
@@ -278,37 +278,41 @@ def storage(self):
...
+                                set(data.get("rgw_configurations")))):
                             role_rgw=True
 
-        if not (role_mds or role_rgw):
+                if not role_ganesha:
+                    role_ganesha = 'ganesha' in data['roles']
+
+        if not (role_mds or role_rgw) and role_ganesha:
             msg = "Ganesha requires either mds or rgw node in cluster."
             self.errors['ganesha'] = msg
```
Comment (supriti) [inline srv/modules/runners/validate.py]: We can set role_ganesha True in the previous loop. 
Maybe rewrite as. 
if role_ganesha:
  if(role_rgw or role_mds):
    self.passed['ganesha'] = "valid"
  else:
     msg = "Ganesha requires either mds or rgw node in cluster."
     self.errors['ganesha'] = msg

Comment (supriti) [inline srv/modules/runners/validate.py]: To me, the code flow seems more intuitive this way. But correctness wise, both approach are ok.

File: srv/modules/runners/validate.py (hunk 1 above)
Comment (jan--f) [inline srv/modules/runners/validate.py]: Yeah I din't quite understand that part. Why does this code binary-and the two sets?

Comment (supriti) [inline srv/modules/runners/validate.py]: User can define custom roles in ganesha_configurations. We are taking intersection of ganesha_configurations and defined roles for node.

**Pull request merged**